### DISCUSSION TITLE
No sunrise on the north pole in 2023?

### DISCUSSION THREAD

**Discussion by bjarkeandersen** [cosinekitty/astronomy#260] [Q&A] [answered]
I'm trying to figure out the next time a sunrise occurs for different places on the Earth. When trying to get the sunrise for the north pole the following code (c#) returns null:

Astronomy.SearchRiseSet(Body.Sun, new Observer(90, 0, 0), Direction.Rise, new AstroTime(DateTime.Parse("2022-07-01")), 600);

if I increase limitDays to 700, I get the date {2024-03-18T00:36:05.238Z}. So my question is: Why isn't the sun rising on the north pole in 2023? If I use the coordinate 89.9, 0  I do get a sunrise in 2023 ({2023-03-19T00:52:44.158Z}). A bug in the code or am I doing something wrong here?

**Comment by cosinekitty** [comment]
You are not doing anything wrong, and it does sound like a bug. I know there might be some iffy behavior near the poles regarding rise/set searches. As one might imagine, there are complications at extreme latitudes.

I will investigate. I suspect I need to add some special case logic to handle things like the Sun just barely rising for the first time after months of darkness, then setting again a few minutes later, etc.

**Reply by cosinekitty** [reply]
Researching this, I went back to the [US Navy rise/set calculator](https://aa.usno.navy.mil/data/RS_OneYear) that I have used in the past to generate test data for sunrise, sunset, moonrise, and moonset. I plugged in the North Pole and found that it has a single sunrise and a single sunset in an entire year. For example, in 2022, the Sun rises on March 18 at 13:02 UTC, and sets September 25 at 04:13 UTC. These are near the equinoxes, which makes sense.

Seeing this gave me a sudden clarity I was lacking before. I can see now why Astronomy Engine is messing this up. If you are at one of the poles, sunrise is no longer a phenomenon caused by the Earth's rotation. Amusingly, the rotation of the Earth is irrelevant for sunrise there. The only thing that matters is the Sun's declination, which is a seasonal effect.

So I'm going to have to re-think rise/set calculations to handle being on or near the poles. The Sun is not the only thing I have to handle. The Moon and planets have more complicated movements through the sky, and my algorithm has to handle all cases somehow. I don't know how long this will take to get sorted, but I will fix it.

In the meantime, the above link may be helpful for what you are doing. I will use it to generate new test data to validate my future fixes.

**Reply by bjarkeandersen** [reply]
Great that you are looking into this so quickly. I found another curiosity for the north pole: When calling SearchHourAngle it returns the min. point in the sky at angle 0/24 and the max. point at angle 12. This happens only when the sun is rising (i.e., from January to June for the north pole). This must be because the sun is rising continuously for half a year and therefore will be higher in the sky at midnight compared to noon for the same day

**Reply by cosinekitty** [reply]
Yes, I was thinking about hour angles near the poles also. It turns out that hour angles are undefined at the poles. For example, hour angle zero is defined as the moment a body crosses the meridian, the semicircle that passes through the observer's sky from north to south. But if you are standing on the north pole, there is no clear direction to pick. Every direction you face is south! I may have to report an error condition if someone passes +90 or -90 as latitude to SearchHourAngle. It's like dividing by zero.

**Reply by cosinekitty** [reply]
An update about hour angles: the current implementation of SearchHourAngle uses the observer's longitude to determine the observer's meridian. If you are standing on one of the poles, the longitude can have any value you want and still refer to the same location. You will get different hour angle results depending on what longitude you choose to assign to the pole.

One way to think of this is in terms of calculus limits. If latitude = +90 and longitude = 0, it is like you have approached the pole from Greenwich England, so you still consider your meridian to be the celestial arc passing through Greenwich and the north pole. You could then rotate your body 180 degrees to face the direction of the zero hour angle, i.e., toward Greenwich. It is the same answer you get if you stopped just one meter before reaching the pole, and you are still standing on the Greenwich side of the pole.

If instead you choose to assign a longitude of -90 to the North Pole, you are arbitrarily selecting New Orleans, Louisiana as your zero hour direction from the north pole.

This is a little bit goofy, but because it is possible to create a meaningful interpretation, I'm inclined now to leave SearchHourAngle as-is, instead of preventing use at the poles with an error result. I will focus on solving the real problem, which is the failure of finding rise/set times reliably inside the arctic/antarctic circles.

**Comment by cosinekitty** [comment]
I am still working on this issue, and I'm making some progress. Now I have formulas for quickly calculating the exact minimum/maximum altitude angles of a body seen over a 24-hour period, given the body's declination and the observer's latitude and elevation. I plan on using these formulas to intelligently adjust the time steps in the rise/set searches, so that I can't possibly skip over a rise/set even in the extreme cases found near the poles. The time steps will be as large as possible to make the search run quickly, but small as needed in the special cases. This is a complicated algorithm, but I will get it working eventually.

**Comment by cosinekitty** [comment, marked as answer]
The rise/set problems have been fixed as of Astronomy Engine v2.1.11, which I just published minutes ago. @bjarkeandersen, would you please try again and let me know how it's working?

**Reply by bjarkeandersen** [reply]
It works perfectly 🥇
